### PR TITLE
ENH: Add ITK helpers for streaming SmartSPIM subimages

### PR DESCRIPTION
**Pull request by tbirdso**
Provides scripts and demonstrates procedure for working with SmartSPIM images in ITK:
1. Compose multiresolution image metadata from `OME-Zarr` standard header and `acquisition.json` bespoke header
2. Retrieve image regions of interest from the cloud with physically correct metadata

Code changes:
- Adds helper scripts in `image.py` and `smartspim.py` for retrieving SmartSPIM physical metadata and translating between physical and voxel domains
- Adds example `StreamSmartSPIM.ipynb` notebook describing procedure for streaming SmartSPIM images with ITK tools

image

image

### CODE REVIEW

File: src/aind_ccf_alignment_experiments/image.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+def arr_to_continuous_index(arr: npt.ArrayLike) -> itk.ContinuousIndex:
+    assert arr.ndim == 1
+    print(arr)
+    itk_index = itk.ContinuousIndex[itk.D, arr.shape[0]]()
+    for index in range(arr.shape[0]):
+        itk_index.SetElement(index, float(arr[index]))
+    return itk_index
+
+
+def get_sample_bounds(image: itk.Image, transform: itk.Transform = None):
```
Comment (thewtex) [inline src/aind_ccf_alignment_experiments/image.py:18]: very nice

File: src/aind_ccf_alignment_experiments/image.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+###############################################################################
+
+
+def block_to_physical_size(
+    block_size: npt.ArrayLike,
+    ref_image: itk.Image,
+    transform: itk.Transform = None,
+) -> npt.ArrayLike:
+    """Convert from voxel block size to corresponding size in physical space"""
+    block_size_tuple = [int(x) for x in block_size]
```
Comment (thewtex) [inline src/aind_ccf_alignment_experiments/image.py]: for clarity, we could have

```
block_size_tuple = tuple(int(x) for x in block_size)
```

i.e., the `_tuple` is a `tuple`.

or

```
index = [int(x) for x in block_size]
```

(we identify the `index` as an index and avoid naming conflicts with the `block_size` parameter).

File: src/aind_ccf_alignment_experiments/smartspim.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+# By convention ITK operations take place in
+# right-to-left, anterior-to-posterior, inferior-to-superior (LPS) space.
+ITK_TARGET_SPACE = (
+    itk.SpatialOrientationEnums.ValidCoordinateOrientations_ITK_COORDINATE_ORIENTATION_LPS
+)
+
+# itk.OMEZarrNGFFImageIO (v0.1.7) does not yet perform axis unit conversions.
+# By convention ITK operations typically assume millimeter spacing,
+# whereas AIND SmartSPIM volumes typically have spacings encoded in micrometers.
+# TODO spacing conversion may be unnecessary in later versions of "ITKIOOMEZarrNGFF".
```
Comment (thewtex) [inline src/aind_ccf_alignment_experiments/smartspim.py:36]: Do we have an issue in ITKIOOMEZarrNGFF that we can reference here?

**Review by thewtex — COMMENTED** [review]
@tbirdso beautiful!! :sparkles: :sparkling_heart: 

LGTM! A few minor comments inline.

CC @lydiang

File: src/aind_ccf_alignment_experiments/image.py (hunk 2 above)
Comment (tbirdso) [reply src/aind_ccf_alignment_experiments/image.py]: Good catch, will rename as `block_index` for clarity

File: src/aind_ccf_alignment_experiments/smartspim.py (hunk 3 above)
Comment (tbirdso) [reply src/aind_ccf_alignment_experiments/smartspim.py:36]: Added inline references to https://github.com/InsightSoftwareConsortium/ITKIOOMEZarrNGFF/issues/42

**Review by thewtex — APPROVED** [review]

**Comment by tbirdso** [comment]
Squashed fixups.

**Comment by tbirdso** [comment]
Moving forward with merge to support related development.

**Pull request merged**